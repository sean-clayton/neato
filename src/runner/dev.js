// @flow

import webpack from 'webpack'
import Server from 'webpack-dev-server'
import webpackConfig from '../webpack/webpack'

export const configHmrWebpackEntry = (neatoConfig: NeatoConfigType): Object => {
  const entry = webpackConfig(neatoConfig).entry

  const hmrEntries = [
    `${require.resolve('webpack-dev-server/client/')}?http://0.0.0.0:${neatoConfig.options.port}`,
    'webpack/hot/only-dev-server'
  ]

  const newEntry = () => {
    if (typeof entry === 'object' && !Array.isArray(entry)) {
      return Object.keys(entry).reduce((entries, key) => ({
        ...entries,
        [key]: hmrEntries.concat(entry[key])
      }), {})
    } else {
      return hmrEntries.concat(entry)
    }
  }

  return {
    ...webpackConfig(neatoConfig),
    entry: newEntry()
  }
}

export default (neatoConfig: NeatoConfigType): Promise<*> => new Promise((resolve: Function, reject: Function) => {
  const userDevServerConfig = neatoConfig.options.webpack && neatoConfig.options.webpack.devServer
    ? neatoConfig.options.webpack.devServer
    : {}

  const devServerConfig = {
    inline: true,
    hot: true,
    historyApiFallback: neatoConfig.options.pages && neatoConfig.options.pages[0] && `${neatoConfig.options.pages[0]}.html`,
    stats: {
      colors: true,
      chunks: false,
      noInfo: true
    },
    ...userDevServerConfig
  }

  try {
    new Server(webpack(configHmrWebpackEntry(neatoConfig)), devServerConfig).listen(neatoConfig.options.port, '0.0.0.0', (err: any) => {
      if (err) {
        console.error(`Server failed to start at http://0.0.0.0:${neatoConfig.options.port}`)
        console.error(err)
      } else {
        console.log(`Server started at http://localhost:${neatoConfig.options.port}/webpack-dev-server/`)
      }
    })
  } catch (e) {
    reject(e)
  }
})
