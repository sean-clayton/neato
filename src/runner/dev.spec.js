/* eslint-env jest */

import { configHmrWebpackEntry } from './dev'

describe('dev server', () => {
  it('Generates the correct entry point', () => {
    const neatoConfig = {
      neatoPath: '/test/neato/path',
      projectPath: '/test/project/path',
      options: {
        pages: ['index', 'other-page'],
        port: 3000,
        vendor: ['a', 'b', 'c']
      }
    }
    expect(configHmrWebpackEntry(neatoConfig).entry).toMatchSnapshot()
  })
})
