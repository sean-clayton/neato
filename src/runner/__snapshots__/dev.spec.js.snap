exports[`dev server Generates the correct entry point 1`] = `
Object {
  "index": Array [
    "/home/sean/code/neato/node_modules/webpack-dev-server/client/index.js?http://0.0.0.0:3000",
    "webpack/hot/only-dev-server",
    "./index",
  ],
  "other-page": Array [
    "/home/sean/code/neato/node_modules/webpack-dev-server/client/index.js?http://0.0.0.0:3000",
    "webpack/hot/only-dev-server",
    "./other-page",
  ],
  "vendor": Array [
    "/home/sean/code/neato/node_modules/webpack-dev-server/client/index.js?http://0.0.0.0:3000",
    "webpack/hot/only-dev-server",
    "a",
    "b",
    "c",
  ],
}
`;
